internal/metrics/remote_write.go

```diff
@@ -0,0 +1,394 @@
+	waitInterval, _ := time.ParseDuration(os.Getenv("REMOTEWRITE_WAIT_INTERVAL"))
+	if waitInterval == 0 {
+		waitInterval = defaultWaitInterval
+	}
+	RangeDuration, _ := time.ParseDuration(os.Getenv("REMOTEWRITE_RANGE_DURATION"))
+	if RangeDuration == 0 {
+		RangeDuration = defaultRequestDuration
+	}
+	requestNumSamples, _ := strconv.Atoi(os.Getenv("REMOTEWRITE_REQUEST_NUM_SAMPLES"))
+	if requestNumSamples == 0 {
+		requestNumSamples = defaultRequestNumSamples
+	}
+	requestRetryInterval, _ := time.ParseDuration(os.Getenv("REMOTEWRITE_REQUEST_RETRY_INTERVAL"))
+	if requestRetryInterval == 0 {
+		requestRetryInterval = defaultRequestRetryInterval
+	}
+
+	newRemoteWrite := RemoteWrite{
+		deviceID:             deviceID,
+		tsdb:                 tsdbInstance,
+		waitInterval:         waitInterval,
+		RangeDuration:        RangeDuration,
+		RequestNumSamples:    requestNumSamples,
+		RequestRetryInterval: requestRetryInterval,
+		lastWriteFile:        path.Join(dataDir, LastWriteFileName),
+	}
+
+	lastWriteBytes, err := ioutil.ReadFile(newRemoteWrite.lastWriteFile)
+	if err == nil {
+		if len(lastWriteBytes) != 0 {
+			i, err := strconv.ParseInt(string(lastWriteBytes), 10, 64)
+			if err == nil {
+				newRemoteWrite.LastWrite = time.Unix(i/1000000000, i%1000000000)
+			} else {
+				log.Error(err)
+			}
+		}
+	} else if !os.IsNotExist(err) {
+		log.Error(err)
+	}
+
+	return &newRemoteWrite
+}
+
+// RemoteRecoverableError
+// 'remote.RecoverableError' fields are private and can't be instantiated during testing. Therefore, we use this wrapper
+type RemoteRecoverableError struct {
+	error
+}
+
+func (e RemoteRecoverableError) Error() string {
+	if e.error != nil {
+		return e.Error()
+	} else {
+		return ""
+	}
+}
+
+// WriteClient
+// an interface used for testing
+//go:generate mockgen -package=metrics -destination=write_client_mock.go . WriteClient
+type WriteClient interface {
+	Write(context.Context, []byte) error
+}
+
+type writeClient struct {
+	client remote.WriteClient
+}
+
+func (w *writeClient) Write(ctx context.Context, data []byte) error {
+	err := w.client.Store(ctx, data)
+	switch err.(type) {
+	case remote.RecoverableError:
+		return RemoteRecoverableError{err}
+	default:
+		return err
+	}
+}
+
+func (r *RemoteWrite) Update(config models.DeviceConfigurationMessage) error {
+	// the following two variables are only here until we implement reading form device configuration
+	serverURL := os.Getenv("REMOTEWRITE_SERVER_URL")
+	serverTimeout, _ := time.ParseDuration(os.Getenv("REMOTEWRITE_SERVER_TIMEOUT"))
+	if serverTimeout == 0 {
+		serverTimeout = time.Second * 10
+	}
+
+	// create configuration
+	r.config = RemoteWriteConfiguration{
+		Url:     serverURL,
+		Timeout: serverTimeout,
+	}
+
+	if r.config.Url == "" { // feature is disabled
+		return nil
+	}
+
+	// create client
+	if r.Client == nil {
+		serverURL, err := url.Parse(r.config.Url)
+		if err != nil {
+			return err
```

Comment:
adding a log here would be helpful during troubleshooting

Reply:
the caller will log the error.

Reply:
fixed in previous changes